lib/web_ui/lib/src/engine/ulps.dart

```diff
@@ -0,0 +1,194 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+// This is a small library to handle stability for floating point operations.
+//
+// Since we are representing an infinite number of real numbers in finite
+// number of bits, when we perform comparisons of coordinates for paths for
+// example, we want to make sure that line and curve sections that are too
+// close to each other (number of floating point numbers
+// representable in bits between two numbers) are handled correctly and
+// don't cause algorithms to fail when we perform operations such as
+// subtraction or between checks.
+//
+// Small introduction into floating point comparison:
+//
+// For some good articles on the topic, see
+// https://randomascii.wordpress.com/category/floating-point/page/2/
+//
+// Here is the 32 bit IEEE representation:
+//   uint32_t mantissa : 23;
+//   uint32_t exponent : 8;
+//   uint32_t sign : 1;
+// As you can see it was carefully designed to be reinterpreted as an integer.
+//
+// Ulps stands for unit in the last place. ulp(x) is the gap between two
+// floating point numbers nearest x.
+
+/// Converts a sign-bit int (float interpreted as int) into a 2s complement
+/// int. Also converts 0x80000000 to 0. Allows result to be compared using
+/// int comparison.
+int signBitTo2sCompliment(int x) => (x & 0x80000000) != 0 ? (-(x & 0x7fffffff)) : x;
```

Comment:
compliment => complement (applies throughout the file)

Reply:
Done.